File: py-server/routes/ico.py
"""
ICO conversion route for Pixora Python backend.
POST /api/ico/convert  — Returns ZIP file with favicon.ico + individual PNGs
"""
from fastapi import APIRouter, UploadFile, File, HTTPException
from fastapi.responses import Response

from services.ico_converter import build_ico_bytes, build_individual_pngs
from services.zip_generator import build_ico_zip
from utils.formats import normalize_format, ICO_INPUT_FORMATS

router = APIRouter()


@router.post("/convert")
async def convert_to_ico(file: UploadFile = File(...)):
    ext = (file.filename or "upload").rsplit(".", 1)[-1]
    fmt = normalize_format(ext)

    if fmt not in ICO_INPUT_FORMATS:
        raise HTTPException(
            status_code=400,
            detail=f"Unsupported input format for ICO conversion: {fmt}. ICO files cannot be converted to ICO.",
        )

    try:
        data = await file.read()

        ico_bytes = build_ico_bytes(data, fmt)
        png_files = build_individual_pngs(data, fmt)
        zip_bytes = build_ico_zip(ico_bytes, png_files)

        headers = {
            "X-Output-Size": str(len(zip_bytes)),
            "Content-Disposition": 'attachment; filename="pixora_favicon_package.zip"',
        }
        return Response(
            content=zip_bytes,
            media_type="application/zip",
            headers=headers,
        )

    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: src-tauri/src/commands/ico.rs
/// POST /api/ico/convert
/// Converts any supported image into a ZIP favicon package.
/// Returns application/zip with favicon.ico + 6 individual PNGs.

use axum::{
    extract::Multipart,
    http::{StatusCode, HeaderValue, header},
    response::{Response, IntoResponse},
};
use serde_json::json;

use crate::services::ico_converter::build_ico_package;
use crate::services::zip_generator::build_zip;
use crate::utils::formats::{ico_input_formats, ext_from_filename};

pub async fn handle_ico_convert(mut multipart: Multipart) -> Response {
    let mut file_data: Option<Vec<u8>> = None;
    let mut filename:  String          = "upload".into();

    while let Ok(Some(field)) = multipart.next_field().await {
        if field.name().unwrap_or("") == "file" {
            filename  = field.file_name().unwrap_or("upload").to_string();
            let bytes = field.bytes().await.unwrap_or_default();
            file_data = Some(bytes.to_vec());
            break;
        }
    }

    let data = match file_data {
        Some(d) => d,
        None    => return err_resp(StatusCode::BAD_REQUEST, "No file field found"),
    };

    // Reject ICO input and unsupported formats
    let ext = ext_from_filename(&filename);
    if ext == "ico" {
        return err_resp(
            StatusCode::BAD_REQUEST,
            "ICO files cannot be converted to ICO. Upload a PNG, JPG, WEBP, etc.",
        );
    }
    if !ico_input_formats().contains(ext.as_str()) {
        return err_resp(
            StatusCode::BAD_REQUEST,
            &format!("Unsupported input format for ICO conversion: {}", ext),
        );
    }

    // Build ICO package
    let package = match build_ico_package(&data, &filename) {
        Ok(p)  => p,
        Err(e) => return err_resp(StatusCode::INTERNAL_SERVER_ERROR, &e),
    };

    // Build ZIP
    let zip_bytes = match build_zip(&package) {
        Ok(z)  => z,
        Err(e) => return err_resp(StatusCode::INTERNAL_SERVER_ERROR, &e),
    };

    let mut headers = axum::http::HeaderMap::new();
    headers.insert(
        header::CONTENT_TYPE,
        HeaderValue::from_static("application/zip"),
    );
    headers.insert(
        header::CONTENT_DISPOSITION,
        HeaderValue::from_static(r#"attachment; filename="pixora_favicon_package.zip""#),
    );
    headers.insert(
        "X-Output-Size",
        HeaderValue::from_str(&zip_bytes.len().to_string())
            .unwrap_or(HeaderValue::from_static("0")),
    );
    headers.insert(
        "Access-Control-Expose-Headers",
        HeaderValue::from_static("X-Output-Size"),
    );

    (StatusCode::OK, headers, axum::body::Body::from(zip_bytes)).into_response()
}

fn err_resp(status: StatusCode, msg: &str) -> Response {
    (status, axum::Json(json!({ "detail": msg }))).into_response()
}